faithcomesbyhearing/dbp
# FCBHDBP-515 enable non-integer verse identifier

# Description
The verse_start column has been changed from an integer type to a string type because sometimes a verse is represented as a string, such as '3a'. As a result, we have updated the verse_start column for the following entities: `bible_verses`, `bible_files`, `bible_file_timestamps`, `user_bookmarks`, `user_highlights`, and `user_notes`. Then, we have created a new column called `verse_sequence` of type integer for each above entity. As a result, we have replaced the column `verse_start` with `verse_sequence` in order to sort the records.

# NOTE 1
We need to run the following SQL statements to create the new column for all entities.

```sql
-- bible_verses
ALTER TABLE bible_verses add column verse_sequence tinyint unsigned NOT NULL;

UPDATE bible_verses SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE bible_verses DROP KEY unique_text_reference;

ALTER TABLE bible_verses modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;

ALTER TABLE bible_verses ADD UNIQUE KEY `unique_text_reference` (`hash_id`,`book_id`,`chapter`,`verse_start`);
------------------------------------------------------------------------------------------------------------------------------------------------

-- bible_file_timestamps
ALTER TABLE bible_file_timestamps add column verse_sequence tinyint unsigned DEFAULT NULL;

UPDATE bible_file_timestamps SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE bible_file_timestamps modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- bible_files
ALTER TABLE bible_files add column verse_sequence tinyint unsigned DEFAULT NULL;

UPDATE bible_files SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE bible_files modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- user_notes
ALTER TABLE dbp_users.user_notes add column verse_sequence tinyint unsigned NOT NULL;

UPDATE dbp_users.user_notes SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE dbp_users.user_notes modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- user_bookmarks
ALTER TABLE dbp_users.user_bookmarks add column verse_sequence tinyint unsigned NOT NULL;

UPDATE dbp_users.user_bookmarks SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE dbp_users.user_bookmarks modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

-- user_highlights
ALTER TABLE dbp_users.user_highlights add column verse_sequence tinyint unsigned NOT NULL;

UPDATE dbp_users.user_highlights SET verse_sequence = verse_start WHERE id > 0;

ALTER TABLE dbp_users.user_highlights modify column verse_start char(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL;
------------------------------------------------------------------------------------------------------------------------------------------------

```

## Issue Link
https://fullstacklabs.atlassian.net/browse/FCBHDBP-515

## How Do I QA This

We should run the following postman tests and they should pass:
- [M] bible.is mobile requests
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/collection/12519377-87d1d0ee-d642-4fb7-ba80-606638d4e5e3?action=share&creator=17122915

- [M] Download requests
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/collection/12519377-3a68ded7-3a3d-40a3-9feb-93ce491311a8?action=share&creator=17122915

- Annotations tests
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/folder/12519377-3cc08c6e-6640-45c3-bc82-96d2d51162d3

- Notes
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-6ebd415f-3060-422b-8b9a-dcd5ee675028
